### PR TITLE
Generate RegionBackendService in terraform

### PR DESCRIPTION
**Pull request by slevenick**
Generate RegionBackendService via MM for terraform. There should be no functionality change in the resulting resource. 

-----------------------------------------------------------------
# [all]

### CODE REVIEW

File: products/compute/api.yaml [hunk 1]
```diff
@@ -935,6 +935,266 @@ objects:
...
+              existing Cache-Control header. The actual headers served in
+              responses will not be altered.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'connectionDraining'
+        description: |
+          Settings for connection draining
+        properties:
+          - !ruby/object:Api::Type::Integer
+            name: 'drainingTimeoutSec'
+            default_value: 0
```
Comment (slevenick) [inline products/compute/api.yaml]: This defaults to 0 to keep with current functionality

File: products/compute/api.yaml [hunk 2]
```diff
@@ -935,6 +935,266 @@ objects:
...
+          the regular expression `[a-z]([-a-z0-9]*[a-z0-9])?` which means the
+          first character must be a lowercase letter, and all following
+          characters must be a dash, lowercase letter, or digit, except the last
+          character, which cannot be a dash.
+      - !ruby/object:Api::Type::Enum
+        name: 'protocol'
+        description: |
+          The protocol this BackendService uses to communicate with backends.
+
+          The possible values are TCP and UDP, and the default is TCP.
```
Comment (slevenick) [inline products/compute/api.yaml]: Do we prefer to allow the API to default this or default via `default_value`?

File: products/compute/terraform.yaml [hunk 3]
```diff
@@ -233,6 +233,68 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         diff_suppress_func: 'compareSelfLinkOrResourceName'
       timeoutSec: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
+  RegionBackendService: !ruby/object:Overrides::Terraform::ResourceOverride
```
Comment (slevenick) [inline products/compute/terraform.yaml]: This is duplicated from the above `BackendService` as they share a nearly identical structure

File: third_party/terraform/tests/resource_compute_region_backend_service_test.go.erb [hunk 4]
```diff
@@ -210,24 +210,6 @@ func TestAccComputeRegionBackendService_withSessionAffinity(t *testing.T) {
 	}
 }
 
-func testAccCheckComputeRegionBackendServiceDestroy(s *terraform.State) error {
```
Comment (slevenick) [inline third_party/terraform/tests/resource_compute_region_backend_service_test.go.erb]: Uses the generated destroy function now

File: products/compute/api.yaml [hunk 5]
```diff
@@ -935,6 +935,266 @@ objects:
...
+
+                Note that you must specify an Instance Group
+                resource using the fully-qualified URL, rather than a
+                partial URL.
+
+                When the BackendService has load balancing scheme INTERNAL, the
+                instance group must be within the same region as the
+                BackendService.
+      - !ruby/object:Api::Type::Enum
+        name: 'loadBalancingScheme'
```
Comment (slevenick) [inline products/compute/api.yaml]: This needs to be included in the API call, but must always be `INTERNAL`. Is there a better way to handle this?

File: products/compute/terraform.yaml (hunk 3 above)
Comment (rileykarson) [reply products/compute/terraform.yaml]: It's my experience that this is often untrue. While load balancing Regional / Global resources will share a common body, each will use only around 60-70% of the fields.

File: products/compute/api.yaml (hunk 1 above)
Comment (rileykarson) [reply products/compute/api.yaml]: We can do this as an override in TF so Ansible / Inspec infer correct values later on if you'd like as well.

File: products/compute/api.yaml (hunk 2 above)
Comment (rileykarson) [reply products/compute/api.yaml]: We should set a default_value and then not use it in Terraform I think- adding one probably counts as a breaking change.

File: products/compute/api.yaml (hunk 5 above)
Comment (rileykarson) [reply products/compute/api.yaml]: Omitting it entirely or using an encoder. This is fine though.

File: products/compute/api.yaml (hunk 2 above)
Comment (slevenick) [reply products/compute/api.yaml]: Is it possible to override a default_value to remove it?

Comment (rileykarson) [reply products/compute/api.yaml]: No idea! If it's impossible, I'd comment out the `default_value` in the api.yaml and note that downstreams need to implement it as an override

File: products/compute/api.yaml [hunk 6]
```diff
@@ -935,6 +935,154 @@ objects:
...
+          Can be NONE, CLIENT_IP, CLIENT_IP_PROTO, or CLIENT_IP_PORT_PROTO.
+
+          When the protocol is UDP, this field is not used.
+        values:
+          - :NONE
+          - :CLIENT_IP
+          - :CLIENT_IP_PROTO
+          - :CLIENT_IP_PORT_PROTO
+      - !ruby/object:Api::Type::ResourceRef
+        name: 'region'
```
Comment (rileykarson) [inline products/compute/api.yaml]: This needs to have an override for `default_from_api` so it's `Computed` + `Optional`

File: products/compute/terraform.yaml [hunk 7]
```diff
@@ -233,6 +233,46 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "region_backend_service_basic"
+        primary_resource_id: "default"
+        vars:
+          region_backend_service_name: "region-backend-service"
+          health_check_name: "health-check"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
```
Comment (rileykarson) [inline products/compute/terraform.yaml]: I think most of this shouldn't be here w/o `iap` and `security_policy` right?

File: products/compute/api.yaml (hunk 6 above)
Comment (slevenick) [reply products/compute/api.yaml]: Why would region be computed and optional? It's required for a regional backend service

File: products/compute/terraform.yaml (hunk 7 above)
Comment (slevenick) [reply products/compute/terraform.yaml]: Right! Removing

**Comment by slevenick** [comment]
> I think you need to delete the handwritten docs still. For some reason the duplicate generation checking doesn't pick it up, probably because it's copied and not compiled.
> 
> Also, there's a few tests we could make redundant; if https://github.com/terraform-providers/terraform-provider-google-beta/blob/c5816035a0cd37713301bf0edab249ab2a419bd2/google-beta/resource_compute_region_backend_service_test.go#L123-L148 is the example, we can remove that handwritten case entirely. (Old checks against specific fields were used before import-style tests, so they can be ignored generally)

Removed a couple tests, deleted docs. I'm inclined to leave the rest of the tests as they include updates

File: products/compute/api.yaml (hunk 6 above)
Comment (rileykarson) [reply products/compute/api.yaml]: We can pull the region from a few different places, so it ends up being optional: https://github.com/terraform-providers/terraform-provider-google/blob/5ce5686d4997d4b422e63f5e295bc9763f2a997d/google/utils.go#L39-L46

File: products/compute/api.yaml [hunk 8]
```diff
@@ -935,6 +935,154 @@ objects:
         description: |
           How many seconds to wait for the backend before considering it a
           failed request. Default is 30 seconds. Valid range is [1, 86400].
+  - !ruby/object:Api::Resource
+    name: 'RegionBackendService'
```
Comment (rileykarson) [inline products/compute/api.yaml]: Let's add [`references`](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/products/compute/api.yaml#L57-L61) to preserve docs like https://github.com/terraform-providers/terraform-provider-google-beta/pull/645/files#diff-83aa3ab4e2ca346150e6962897976dbaL12 for readers

File: products/compute/api.yaml [hunk 9]
```diff
@@ -935,6 +935,154 @@ objects:
...
+    kind: 'compute#backendService'
+    base_url: projects/{{project}}/regions/{{region}}/backendServices
+    collection_url_response: !ruby/object:Api::Resource::ResponseList
+      kind: 'compute#backendServicesList'
+      items: 'items'
+    has_self_link: true
+    description: |
+      A Region Backend Service defines a regionally-scoped group of virtual
```
Comment (rileykarson) [inline products/compute/api.yaml]: It's probably worth noting that RBS is only for ILB, like we did before: https://github.com/terraform-providers/terraform-provider-google-beta/pull/645/files#diff-83aa3ab4e2ca346150e6962897976dbaL15

File: products/compute/api.yaml (hunk 6 above)
Comment (slevenick) [reply products/compute/api.yaml]: Like this?
```
      region: !ruby/object:Overrides::Terraform::PropertyOverride
        default_from_api: true
        required: false
```

File: products/compute/api.yaml [hunk 10]
```diff
@@ -935,6 +935,162 @@ objects:
...
+      machines that will serve traffic for load balancing.
+
+      ~> **Note**: Region backend services can only be used when using internal load balancing.
+      For external load balancing, use
+      [`google_compute_backend_service`](compute_backend_service.html) instead. 
+
+
+ ## Example Usage
```
Comment (rileykarson) [inline products/compute/api.yaml]:
```suggestion
 ## Example Usage
```

This seems unintentional

File: products/compute/api.yaml [hunk 11]
```diff
@@ -935,6 +935,162 @@ objects:
...
+    base_url: projects/{{project}}/regions/{{region}}/backendServices
+    collection_url_response: !ruby/object:Api::Resource::ResponseList
+      kind: 'compute#backendServicesList'
+      items: 'items'
+    has_self_link: true
+    description: |
+      A Region Backend Service defines a regionally-scoped group of virtual
+      machines that will serve traffic for load balancing.
+
+      ~> **Note**: Region backend services can only be used when using internal load balancing.
```
Comment (rileykarson) [inline products/compute/api.yaml]:
```suggestion
      Region backend services can only be used when using internal load balancing.
```

The warning box syntax will appear as symbols in most markdown processors, so let's remove it.

File: products/compute/api.yaml (hunk 6 above)
Comment (rileykarson) [reply products/compute/api.yaml]: Yep! There are some other extra overrides too that make it work a little better as well, see https://github.com/GoogleCloudPlatform/magic-modules/blob/16e6fe42ebdb772e323e3fab7dad503e1de8ea8d/products/compute/terraform.yaml#L52-L58

File: products/compute/api.yaml [hunk 12]
```diff
@@ -935,6 +935,159 @@ objects:
...
+      kind: 'compute#backendServicesList'
+      items: 'items'
+    has_self_link: true
+    description: |
+      A Region Backend Service defines a regionally-scoped group of virtual
+      machines that will serve traffic for load balancing.
+
+      Region backend services can only be used when using internal load balancing.
+      For external load balancing, use
+      [`google_compute_backend_service`](compute_backend_service.html) instead.
```
Comment (rileykarson) [inline products/compute/api.yaml]:
```suggestion
      a global backend service instead.
```

**Pull request merged**